src/elf/symver.rs

```diff
@@ -0,0 +1,902 @@
+    pub struct VersymSection<'a> {
+        bytes: &'a [u8],
+        ctx: container::Ctx,
+    }
+
+    impl<'a> VersymSection<'a> {
+        pub fn parse(
+            bytes: &'a [u8],
+            shdrs: &'_ [SectionHeader],
+            ctx: container::Ctx,
+        ) -> Result<Option<VersymSection<'a>>> {
+            // Get fields needed from optional `symbol version` section.
+            let (offset, size) =
+                if let Some(shdr) = shdrs.iter().find(|shdr| shdr.sh_type == SHT_GNU_VERSYM) {
+                    (shdr.sh_offset as usize, shdr.sh_size as usize)
+                } else {
+                    return Ok(None);
+                };
+
+            // Get a slice of bytes of the `version definition` section content.
+            let bytes: &'a [u8] = bytes.pread_with(offset, size)?;
+
+            Ok(Some(VersymSection { bytes, ctx }))
+        }
+
+        /// Get an iterator over the [`Versym`] entries.
+        #[inline]
+        pub fn iter(&'a self) -> VersymIter<'a> {
+            self.into_iter()
+        }
+
+        /// True if there are no [`Versym`] entries.
+        #[inline]
+        pub fn is_empty(&self) -> bool {
+            self.bytes.is_empty()
+        }
+
+        /// Number of [`Versym`] entries.
+        #[inline]
+        pub fn len(&self) -> usize {
+            let entsize = core::mem::size_of::<ElfVersym>();
+
+            self.bytes.len() / entsize
+        }
+
+        /// Get [`Versym`] entry at index.
+        #[inline]
+        pub fn get_at(&self, idx: usize) -> Option<Versym> {
+            let entsize = core::mem::size_of::<ElfVersym>();
+            let offset = idx.checked_mul(entsize)?;
+
+            self.bytes
+                .pread_with::<ElfVersym>(offset, self.ctx.le)
+                .ok()
+                .map(|vs| Versym {
+                    vs_val: vs.vs_val as usize,
+                })
+        }
+    }
+
+    impl<'a> IntoIterator for &'_ VersymSection<'a> {
+        type Item = <VersymIter<'a> as Iterator>::Item;
+        type IntoIter = VersymIter<'a>;
+
+        fn into_iter(self) -> Self::IntoIter {
+            VersymIter {
+                bytes: self.bytes,
+                offset: 0,
+                ctx: self.ctx,
+            }
+        }
+    }
+
+    /// Iterator over the [`Versym`] entries from the [`SHT_GNU_VERSYM`] section.
+    pub struct VersymIter<'a> {
+        bytes: &'a [u8],
+        offset: usize,
+        ctx: container::Ctx,
+    }
+
+    impl<'a> Iterator for VersymIter<'a> {
+        type Item = Versym;
+
+        fn next(&mut self) -> Option<Self::Item> {
+            if self.offset >= self.bytes.len() {
+                None
+            } else {
+                // Safe to unwrap as the length of the byte slice was validated in VersymSection::parse.
+                let ElfVersym { vs_val } = self
+                    .bytes
+                    .gread_with(&mut self.offset, self.ctx.le)
+                    .unwrap();
```

Comment:
What if `bytes.len()` is not a multiple of `core::mem::size_of::<ElfVersym>()`?

Reply:
Good point, I'll validate the length of the byte slice in `VersymSection::parse`.